corehq/blobs/management/commands/wipe_blobdb.py

```diff
@@ -0,0 +1,70 @@
+import asyncio
+
+from django.core.management import BaseCommand
+
+from corehq.apps.cleanup.utils import confirm_destructive_operation
+from corehq.blobs import get_blob_db
+from corehq.blobs.models import BlobMeta
+from corehq.sql_db.util import get_db_aliases_for_partitioned_query
+from dimagi.utils.chunked import chunked
+
+CHUNK_SIZE = 1000
+
+
+class Command(BaseCommand):
+    """
+    Wipe all data from BlobDB.
+    """
+    def add_arguments(self, parser):
+        parser.add_argument(
+            '--commit',
+            action='store_true',
+            dest='commit',
+            default=False,
+        )
+
+    def handle(self, *args, **options):
+        confirm_destructive_operation()
+
+        # Python 3.7+
+        # bytes_deleted = asyncio.run(wipe_blobdb(options['commit']))
+
+        # Python 3.4+
+        loop = asyncio.get_event_loop()
+        try:
+            future = wipe_blobdb(options['commit'])
+            bytes_deleted = loop.run_until_complete(future)
+        finally:
+            loop.close()
+
+        print(f"Deleted {bytes_deleted} bytes.")
+        if not options['commit']:
+            print("You need to run with --commit for the deletion to happen.")
+
+
+async def wipe_blobdb(commit=False):
+    """
+    Wipe shards in parallel
+    """
+    coros = [wipe_shard(dbname, commit)
+             for dbname in get_db_aliases_for_partitioned_query()]
+    bytes_deleted_list = await asyncio.gather(*coros)
+    return sum(bytes_deleted_list)
+
+
+async def wipe_shard(dbname, commit=False):
+    bytes_deleted = 0
+    metas = BlobMeta.objects.using(dbname).all()
```

Comment:
This could be a really huge result. It will probably be better to use a pagination strategy. Ideally it would do the SUM in a single query along with the DELETE in Postgres. Something like

```sql
WITH deleted AS (
    DELETE FROM blobs_blobmeta
    WHERE id IN (
        SELECT id
        FROM blobs_blobmeta
        ORDER BY id
        LIMIT 1000
    )
    RETURNING *
)
SELECT SUM(content_length) AS bytes_deleted
FROM deleted;
```

Reply:
Isn't it paginated? I thought `all()` is lazy, and `chunked()` is only fetching 1000 rows at a time. ... Or am I misunderstanding something?

Reply:
I don't think `all()` is lazy, or at least it loads all of the results immediately when the first one is accessed.

Reply:
Ah! You are right. [QuerySets are evaluated](https://docs.djangoproject.com/en/1.11/ref/models/querysets/#when-querysets-are-evaluated) when they are iterated, or when they are sliced (among other things).

Django's `Paginator` class [uses slicing](https://github.com/django/django/blob/master/django/core/paginator.py#L78) to evaluate them late ... but should not be used on [large QuerySets](https://docs.djangoproject.com/en/3.0/ref/paginator/#django.core.paginator.Paginator.object_list), so I won't use it here.

I'll change this to use slicing to iterate chunks.

Reply:
> I'll change this to use slicing to iterate chunks.

I think even Django pagination logic will not work well for very large result sets because it uses limit with *offset* (`self.object_list[bottom:top]`), which does not perform well for large offsets.

[Posgres docs](https://www.postgresql.org/docs/8.3/queries-limit.html) say:
> The rows skipped by an `OFFSET` clause still have to be computed inside the server; therefore a large `OFFSET` might be inefficient.

The same issue is referenced in this article https://www.citusdata.com/blog/2016/03/30/five-ways-to-paginate/

Reply:
The Postgres docs explain it better (for my brain at least) than the Django docs. :facepalm: I've just done the same thing, only shorter.

[Disqus solved this](http://cra.mr/2011/03/08/building-cursors-for-the-disqus-api) by using the primary key instead of offset. `BlobMeta`'s primary key is a normal auto-incrementing integer field called `id`, so a paginator can use that.